File: mergesort.c


#include <stdio.h>

#define MAXSIZE 10

typedef struct
{
    int data[MAXSIZE];
    int length;
}SqList;
//排序
void Merge(int SR[], int TR[], int i, int m,int n)
{
    int j,k,l;
    for(j = m + 1,k = i; i <= m && j <= n; k++)
    {
        if (SR[i] < SR[j])
            TR[k] = SR[i++];
        else
            TR[k] = SR[j++];
    }
    if (i <= m)
    {
        for(l = 0; l <= m - i;l++)
        {
            TR[k+l] = SR[i+l];
        }
    }

    if (j <= n)
    {
        for (l = 0;l <= n-j;l++)
        {
            TR[k+l] = SR[j + l];
        }
    }
}
//递归拆分数组，排序
void MSort(int SR[], int TR1[], int s , int t)
{
    int m;


    int TR2[MAXSIZE + 1];
    //递归结束条件
    if (s == t)
    {
        TR1[s] = SR[s];
    }
    else 
    {
        m = (s + t) / 2;
        //递归拆分数组左半部分  
        MSort(SR, TR2, s, m);
        //递归拆分右半部分
        MSort(SR, TR2, m + 1, t);
        //归并
        Merge(TR2, TR1, s, m, t);
  
    } 
}

void MergeSort(SqList *l)
{
    MSort(l->data,l->data,1,4);
}


void main()
 {
    SqList l;
    SqList *ll;
    ll = &l;
    ll->length = 4;
    ll->data[1] = 50;
    ll->data[2] = 10;
    ll->data[3] = 90;
    ll->data[4] = 30;

    MergeSort(ll);


}

File: mergesort.php
<?php

class merge
{
    function mergesort()
    {
        $arr1 = array(
            1 => 50, 2 => 10, 3 => 30,4 => 20
        );
        testf($arr1, $arr1, 1, 4);
    }
}
function testf(&$sr, &$tr1, $s, $t)
{
    $tr2 = array();
    if ($s == $t)
    {
        $tr1[$s] = $sr[$s];
    }else {
        $m = intval(($s + $t) / 2);
        testf($sr, $tr2, $s, $m);
        testf($sr, $tr2, $m + 1, $t);
        echo $s;
        echo "<br>";
        echo $m;
        echo "<br>";
        echo $t;

        var_dump($tr2);
        var_dump($tr1);
        merge($tr2,$tr1,$s,$m,$t);
    }
}
function merge(&$sr, &$tr, $i, $m, $n)
{
    for($j = $m + 1, $k = $i; $i <= $m && $j <= $n; $k++)
    {
        if ($sr[$i] < $sr[$j]) {
            $tr[$k] = $sr[$i++];
        }else {
            $tr[$k] = $sr[$j++];
        }
    }

    if ($i <= $m)
    {
        for($l = 0; $l <= $m- $i; $l++)
        {
            $tr[$k+$l] = $sr[$i + $l];
        }
    }
    if ($j <= $n)
    {
        for ($l = 0; $l <= $n - $j;$l++)
        {
            $tr[$k + $l] = $sr[$j + $l];
        }
    }
}
